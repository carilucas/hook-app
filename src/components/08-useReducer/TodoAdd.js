import React, { useState } from 'react';
import { useForm } from '../../hooks/useForm';   

export const TodoAdd = ({ handleAddTodo }) => {

   const [error, setError] = useState(false);

   const [{description}, handleInputChange, reset] =  useForm({description:''});

   const handleSubmit = e=>{
      e.preventDefault();

      if (description.trim().length <= 3) {
         setError(true);
         return;
      }else{
         setError(false)
      }

      const newTodo = {
         id : new Date().getTime(),
         desc: description,
         done:false
      }

      handleAddTodo(newTodo);
      reset();
   }
   return (
      <>
         <h4>Agregar tarea</h4>
         <hr/>
         <form action="" onSubmit={handleSubmit}>
            <input 
               type="text"
               name="description"
               placeholder="Aprender"
               autoComplete="off"
               className="form-control"
               onChange={handleInputChange}
               value={description}
            />
            {
               error && <div className="alert alert-danger" role="alert">
               Ingresa una descripción para tu taréa de al menos 4 caractéres
               </div>}
            <button
               type="submit"
               className="btn btn-outline-primary mt-1 w-100"
            >
               Agregar
            </button>
         </form>
      </>
   )
}
