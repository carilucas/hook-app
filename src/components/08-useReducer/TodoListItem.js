import React from 'react'

export const TodoListItem = ({todo, index, handleCompleted, handleRemove}) => {
   return (
      <li
         key={todo.id}
         className="list-group-item"
      >
         <p
            className={`${todo.done && "complete"}`}
            onClick={()=> handleCompleted(todo.id)}
         >
            {index + 1}. {todo.desc}
         </p>
         <button
            onClick={()=>handleRemove(todo.id)}
            className="btn btn-danger"
         >
            Borrar
         </button>
      </li>
   )
}
