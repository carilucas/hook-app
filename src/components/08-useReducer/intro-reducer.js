
const initialState = [
   {
      id: 1,
      todo:'Comprar pan',
      done : false
   }
]

const todoReducer = (state=initialState,action)=>{
   if (action && action.type==='AGREGAR') {
      return [...state,action.payload]
   }
   return state
}

let todos = todoReducer();

const todo2 = {
   id: 2,
   todo:'Comprar huevos',
   done : false
}

const agregarTodo = {
   type:'AGREGAR',
   payload: todo2
}

todos = todoReducer(todos,agregarTodo);


console.log(todos);