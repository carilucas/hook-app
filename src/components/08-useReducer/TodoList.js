import React from 'react'
import {TodoListItem} from './TodoListItem'

export const TodoList = ({todos, handleRemove, handleCompleted}) => {
   return (
      <ul className="list-group list-group-flush">
         {
            todos.map((todo,i) => (
               <TodoListItem 
               key={todo.id}
               todo={todo} 
               index={i} 
               handleRemove={handleRemove} 
               handleCompleted={handleCompleted}
               />
               
            ))
         }
      </ul>
   )
}
