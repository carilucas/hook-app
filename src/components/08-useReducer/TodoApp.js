import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import './style.css';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';


const init = ()=>{
   return JSON.parse(localStorage.getItem('Todos')) || []
}

export const TodoApp = () => {

   const [todos,dispatch] = useReducer(todoReducer, [],init);
   
   useEffect(() => {
      localStorage.setItem('Todos' , JSON.stringify(todos));
   }, [todos])

   const handleRemove = (id)=>{
      dispatch({
         type: 'remove',
         payload: id
      })
   }

   const handleCompleted = (id)=>{
      dispatch({
         type:'completed',
         payload: id
      });
   }

   const handleAddTodo = (newTodo)=>{
      dispatch({
         type:'add',
         payload: newTodo
      });  
   }

   return (
      <div>
         <h1>TodoApp: {todos.length}</h1>
         <hr/>
         <div className="row">
            <div className="col-7">
               <TodoList 
                  todos={todos} 
                  handleRemove={handleRemove} 
                  handleCompleted={handleCompleted} 
               />
            </div>
            <div className="col-5">
               <TodoAdd 
                  handleAddTodo={handleAddTodo}
               />
            </div>
         </div>
      </div>
   )
}
