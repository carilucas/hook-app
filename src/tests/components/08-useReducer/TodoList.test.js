import { shallow } from 'enzyme';
import React from 'react';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from '../../fixtures/demoTodos';



describe('Pruebas en <TodoList />', () => {
   const handleRemove = jest.fn();
   const handleCompleted = jest.fn();

   const wrapper = shallow(
      <TodoList 
         todos={demoTodos} 
         handleRemove={ handleRemove } 
         handleCompleted={ handleCompleted }
      />
   );
   
   test('Debe mostrarse correctamente snapshot', () => {
      expect(wrapper).toMatchSnapshot();
   });

   test('Debe mostrar el nÃºmero correcto de los todos', () => {
      
      expect(wrapper.find('TodoListItem').length).toBe(demoTodos.length);
      expect(wrapper.find('TodoListItem').at(0).prop('handleRemove')).toEqual(expect.any(Function));
      expect(wrapper.find('TodoListItem').at(0).prop('handleCompleted')).toEqual(expect.any(Function));

   })
   
   
})
