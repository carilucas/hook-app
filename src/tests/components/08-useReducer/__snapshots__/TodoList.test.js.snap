// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoList /> Debe mostrarse correctamente snapshot 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleCompleted={[MockFunction]}
    handleRemove={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "Descripcion",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleCompleted={[MockFunction]}
    handleRemove={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "Descripcion 2",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
